### DISCUSSION TITLE
Terminal tab title no change when ssh server

### DISCUSSION THREAD

**Discussion by anhdudeptrai** [elementary/terminal#646] [Q&A]
Hi,
when I ssh the server, the title on the tab terminal doesn't change. 
I'm using elementary os 6.1.
How to fix it?

Thank you!
![Screenshot from 2021-12-30 12-41-18](https://user-images.githubusercontent.com/34654634/147724757-ea9722d9-2e92-4dfa-8ab3-393962307978.png)

**Comment by pklausner** [comment]
Same here. The usual control codes which Bash uses to set the title do work, but only for the overall window.
It looks as if the terminal somehow(tm) grabs the working directory of the topmost shell and shows its basename.
Is there a documentation somewhere about which control codes set that tab title?

**Reply by jeremypw** [reply]
Just to be clear, are we talking about the label in the window headerbar or the label in the tabbar? 

The former is currently the same as the Vte.Terminal property `window-title` if not empty, otherwise the current directory path.  I presume the Vte.Terminal gets the window-title from the shell (see `~./bashrc`). It is a read-only property. This is also bound to the window's `title` property.  The tab label is the basename of the current directory (disambiguated if required).

There is an open issue and PR related to the headerbar title (see #826 and #830) but not I think related to this issue.  There is another issue which may be related - #593. Also #698.

**Reply by jeremypw** [reply]
I have pushed a PR that partially fixes the issue by showing the name of the foreground process in the tab label.  So if you are using `ssh` to connect it will show that name in the label.  There is no obvious way to access what is happening within the foreground process though unless that process itself changes the parent `bash` environment or otherwise changes the window title.

**Reply by pklausner** [reply]
Yes, it's about the distinction of tab vs window title.

I guess it was xterm which introduced the control sequences to set **a** title. As the concept of tabs wasn't event known back then, this meant the window title. Other (all?) tabbed terminal programs solve this by setting the tab title. The window title then is the title of the currently active tab. This is good, because theoretically you can distinguish the content of the tabs. This is bad, because the available space for tab titles is much smaller. Typical titles like `user@full.host.name:~/foo/bar` get truncated and you only see sth like
`/user@x..\ /user@x..\ /user@y..\` - not helpful. My personal shell settings work around this by trying to have the most salient bit at the front of the title.

Elementary terminal's idea to use the current directory basename as tab title works very well for local shells. The workaround to show the program name if it's not a shell helps a wee bit.
For ssh, I would prefer to see the hostname first. For other programs, other defaults may make more sense.

Maybe there is a way to "mark" the most important part of the full title and have that as tab title?
Maybe interpreting the bold / un-bold attribute within the title? Normal terminals ignore it, IIUC.

**Reply by jeremypw** [reply]
Thanks for the suggestions! What we could do is to set the window title to the full command line of any foreground process that is running, which should include any arguments.  At the moment it is still showing the cwd of the underlying shell which isn't very helpful. I am hesitant to try and start extracting bits based on what the foreground process is.

**Reply by jeremypw** [reply]
After more investigation, comparing Xterm on OS7.1 and OS8 it looks like the default Xterm window title on OS8 (Ubuntu 24.04) now shows details of the foreground process anyway so I am not sure we actually have to fix it in Terminal.  I think this is because a different `.bashrc` and `etc/profile.d/vte-2.91.sh` file is shipped which takes care of it, at least as far as the window title is concerned.  We will still have to handle the tab labelling. Maybe we could have an option to edit the tab label in case it does not suit the user in a particular case.

**Comment by pklausner** [comment]
True, xterm's default title is just the program it's running. Typically a shell with a maze of rc scripts and prompt commands which send control codes to overwrite the (window) title. And with most terminals, that's also the tab title.

If you don't care too much about this tab label, then elementary's default is an improvement. I do care, so how could _"option to edit the tab label"_ work? A pattern to extract from the full title? Or from the running program or what?

For myself, I have crude work-around for interactive ssh:
```
ssh() 
{ 
  mkdir -p "$HOME/ssh/$1"
  cd "$HOME/ssh/$1"
  ssh "$@"
  cd -
}
```
Btw: sometimes(tm) elementary's Terminal shows 2 directories, so label turns out as e.g. `ssh/192.168.0.9`

**Reply by jeremypw** [reply]
>  I do care, so how could "option to edit the tab label" work?

Initially just  text entry I should think, that would apply to that particular tab instance, possibly restorable.  Just to allow the user to put a short label that is meaningful to them related to the task being done in that tab.  Anything more complicated would need more thought and discussion (maybe open a discussion thread?)

**Reply by jeremypw** [reply]
> Btw: sometimes(tm) elementary's Terminal shows 2 directories, so label turns out as e.g. ssh/192.168.0.9

That sounds like Terminal is trying to disambiguate the label from another one.  This is normally done for directory paths so that if there are two tabs showing two different folders both `build` then it prepends the parent folder to both labels (or as much of the path needed to disambiguate.

**Reply by jeremypw** [reply]
Once we implement a custom-tab-label concept, we can expand its utility e.g. by having a command-line option to add a tab with a custom label.

**Reply by pklausner** [reply]
The disambiguation feature is neat.

To edit the tab label *manually* won't be used very much. Maybe if the title persists over session, but how would you properly do that?

The new kid on the terminal block Ghosty does this:
- _the_ title goes into tab bar
- window title is the active tab's title
- for local shell ignores the terminal sequences, title always is the full pwd
- for any other command, accept the title control commands

I think the best of both worlds is:
- a local shell tab behaves as now
- anything else makes tab title == window title

**Reply by jeremypw** [reply]
The problem with tab title == window title is that the window title is likely to get truncated and we have no control over that (TabBar is sealed). Unless the TabBar is set to expand the tabs to fill the available space (which is not currently exposed) the tabs have a fixed, homogeneous width and ellipsize at the end.

Custom labels could be saved/restored without too much difficulty, but the first step is to get the concept approved (or not).

**Reply by jeremypw** [reply]
There have been request to specify a custom title for a terminal.  I have proposed a PR that allows this to be done both manually and from the commandline.

**Reply by jeremypw** [reply]
I agree it is only ever going to be a corner case but allows the user to choose what is the most relevant information that will fit in the available space (if the default does not suffice).

**Reply by pklausner** [reply]
> The problem with tab title == window title is that the window title is likely to get truncated

Am I misunderstanding something here? I'd think the window title is much, _much_ wider than the tab title:  
- window title width = window width - 4 buttons
- tab title width = (window width - 2 buttons) / n - 1 close button

Even with n = 1, the default settings give a tab label width of max 30 characters of an 80 fixed characters wide window.

**Reply by jeremypw** [reply]
Sorry, I meant the tab title gets truncated.
